tilavarauspalvelu/integrations/keyless_entry/client.py

```diff
@@ -41,124 +51,374 @@ class PindoraClient(BaseExternalServiceClient):
+
+        response = cls.put(url=url)
+        cls._validate_seasonal_booking_response(
+            response,
+            application_section,
+            action="changing access code for seasonal booking",
+        )
+
         data = cls.response_json(response)
-        return cls._parse_reservation_response(data)
+        return cls._parse_seasonal_booking_response(data)
+
+    ######################
+    # Reservation series #
+    ######################
+
+    @classmethod
+    def get_reservation_series(cls, series: RecurringReservation) -> PindoraReservationSeriesResponse:
+        """Fetch a reservation series from Pindora."""
+        url = cls._build_url(f"reservation-serie/{series.ext_uuid}")
+
+        response = cls.get(url=url)
+        cls._validate_reservation_series_response(
+            response,
+            series,
+            action="fetching reservation series",
+        )
+
+        data = cls.response_json(response)
+        return cls._parse_reservation_series_response(data)
+
+    @classmethod
+    def create_reservation_series(
+        cls,
+        series: RecurringReservation,
+        *,
+        is_active: bool = False,
+    ) -> PindoraReservationSeriesResponse:
+        """Create a new reservation series in Pindora."""
+        url = cls._build_url("reservation-series")
+
+        reservations: list[Reservation] = list(series.reservations.all())
+
+        if not reservations:
+            msg = f"No reservations in for reservation series '{series.ext_uuid}'."
+            raise PindoraAPIError(msg)
+
+        data = PindoraReservationSeriesCreateData(
+            reservation_serie_id=str(series.ext_uuid),
+            reservation_unit_id=str(series.reservation_unit.uuid),
+            series=[
+                PindoraReservationSeriesReservationData(
+                    begin=reservation.begin.astimezone(DEFAULT_TIMEZONE).isoformat(),
+                    end=reservation.end.astimezone(DEFAULT_TIMEZONE).isoformat(),
+                )
+                for reservation in reservations
+            ],
+            is_active=is_active,
+        )
+
+        response = cls.post(url=url, json=data)
+        cls._validate_reservation_series_response(
+            response,
+            series,
+            action="creating reservation series",
+        )
+
+        data = cls.response_json(response)
+        return cls._parse_reservation_series_response(data)
+
+    @classmethod
+    def update_reservation_series(
+        cls,
+        series: RecurringReservation,
+        *,
+        is_active: bool = False,
+    ) -> None:
+        """Create a new reservation series in Pindora."""
+        url = cls._build_url(f"reservation-serie/{series.ext_uuid}")
+
+        reservations: list[Reservation] = list(series.reservations.all())
+
+        if not reservations:
+            msg = f"No reservations in for reservation series '{series.ext_uuid}'."
+            raise PindoraAPIError(msg)
```

Comment:
`PindoraAPIError` seems like a wrong error, since the error isn't from PindoraAPI, but our own system 🤔

Reply:
Added a `PindoraClientError` base class that `PindoraAPIError` subclasses. Also renamed `PindoraAPIConfigurationError` to `PindoraClientConfigurationError` since it is also not about the API but rather our config.